File: 102-binary-tree-level-order-traversal/102-binary-tree-level-order-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function levelOrder(root: TreeNode | null): number[][] {
  
  if(!root)
    return [];
  
  const level_order_traversal = new Array<Array<number>>();
  let q = new Array<TreeNode>();
  
  q.push(root);
  
  while(q.length) {
    const curr_q = new Array<TreeNode>(),
          curr_level = new Array<number>();
    
    for(let i = 0; i < q.length; i++) {
      const curr_item = q[i];

      curr_level.push(curr_item.val);

      if(curr_item.left)
        curr_q.push(curr_item.left)

      if(curr_item.right)
        curr_q.push(curr_item.right);
    }
    
    level_order_traversal.push(curr_level);
    q = curr_q;
  }
  
  return level_order_traversal;
};

File: 102-binary-tree-level-order-traversal/102-binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
      List<List<Integer>> ans = new ArrayList<>();
      Queue<TreeNode> q = new LinkedList<>();

      q.offer(root);

      while (!q.isEmpty()) {
        int N = q.size();
        List<Integer> curr = new ArrayList<>();

        if(root == null)
          return ans;

        for (int i = 0; i < N; i++) {
          TreeNode node = q.poll();
          curr.add(node.val);
          if(node.left != null)
            q.offer(node.left);

          if(node.right != null)
            q.offer(node.right);
        }
        ans.add(curr);
      }

      return ans;
    }
}